core/match_pattern.py

```diff
@@ -62,44 +62,99 @@ def compute_child_table(suf_tab, lcp_tab):
     return child_table
 
 
-def get_child_intervals(i, j, child_tab: list[Child]) -> list[Tuple[int, int]]:
-    intervals = []
+def is_interval_eligible(a: str, i: int, lcp: int, s: str, suf_tab: list[int]) -> bool:
+    idx_to_hit = suf_tab[i]
+
+    return s[idx_to_hit + lcp] == a
+
+
+def get_lcp(i: int, j: int, child_tab: list[Child], lcp_tab: list[int]):
+    assert i > -1 and j <= len(child_tab)
+
+    # root interval
+    if i == 0 and j == len(child_tab):
+        return 0
+    elif len(child_tab) > j > child_tab[j].up > i:
```

Comment:
I find it weird that you write this in decreasing order, but okay.

You have already checked that `j <= len(child_tab)` at the top of the function and if `j == len(child_tab)` in the check before, so you already know that `j < len(child_tab)`. It would suffice to check 

```python
    elif i < child_tab[j].up < j:
```